src/cryptonote_core/blockchain.cpp

```diff
@@ -3651,8 +3651,19 @@ uint64_t Blockchain::get_dynamic_base_fee(uint64_t block_reward, size_t median_b
     div128_64(hi, lo, min_block_weight, &hi, &lo, NULL, NULL);
     div128_64(hi, lo, median_block_weight, &hi, &lo, NULL, NULL);
     assert(hi == 0);
-    lo /= 5;
-    return lo;
+    if (version >= HF_VERSION_2021_SCALING)
+    {
+      lo *= CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V5;
+      lo /= median_block_weight;
+      lo -= lo / 20;
+      lo = cryptonote::round_money(lo, CRYPTONOTE_SCALING_2021_FEE_ROUNDING_PLACES);
+      return lo;
+    }
+    else
+    {
+      lo /= 5;
+      return lo;
+    }
   }
```

Comment:
Technically correct for the existing return value of `get_min_block_weight()`, but will become incorrect if that return value changes.

Fix:

```
    lo = mul128(block_reward, DYNAMIC_FEE_REFERENCE_TRANSACTION_WEIGHT, &hi);
    div128_64(hi, lo, median_block_weight, &hi, &lo, NULL, NULL);

    if (version >= HF_VERSION_2021_SCALING)
    {
      // min_fee_per_byte = 0.95 * block_reward * ref_weight / (fee_median^2)
      // note: fee_median computation changed at hardfork HF_VERSION_2021_SCALING
      div128_64(hi, lo, median_block_weight, &hi, &lo, NULL, NULL);
      assert(hi == 0);
      lo -= lo / 20;
      lo = cryptonote::round_money(lo, CRYPTONOTE_SCALING_2021_FEE_ROUNDING_PLACES);
      return lo;
    }
    else
    {
      // min_fee_per_byte = 0.2 * block_reward * ref_weight / (min_penalty_free_zone * fee_median)
      div128_64(hi, lo, min_block_weight, &hi, &lo, NULL, NULL);
      assert(hi == 0);
      lo /= 5;
      return lo;
    }
  }
```

I can't extend the Github selection zone to higher lines... annoying.

Reply:
The fee update research did not recommend rounding the minimum fee from the node's point of view - only the transaction author. However, I think it is fine to round the node's minimum fee, since it will enforce rounding on nodes to some extent.

Reply:
Removing "lo /= median_block_weight;" is a typo, right ?

Reply:
And the other change seems wrong, since it causes the tests to fail (by much more than just the rounding change from 3 to 2 places) - unless the values from the PDF test cases were wrong in the first place.

Reply:
In your code,

```
div128_64(hi, lo, min_block_weight, &hi, &lo, NULL, NULL);
```

Is canceled out by 

```
lo *= CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V5;
```

Why cancel it at all? My suggestion is to skip these calculations because they are zero sum. I don't know why the tests would fail.

> Removing "lo /= median_block_weight;" is a typo, right ?

It is replaced with this, which should be equivalent.

```
div128_64(hi, lo, median_block_weight, &hi, &lo, NULL, NULL);
```

Reply:
nvm, I had missed another change in the block. Anyway, I've done this except the removal of these two "canceling" lines since it's not obvious they cancel and it's not like an extra div here really slows it down.

Reply:
> I've done this except the removal of these two "canceling" lines since it's not obvious they cancel and it's not like an extra div here really slows it down.

I am confused by this reply. Where in the fee doc does it specify making these computations?

Reply:
The fee doc does not specify those.